File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/common/ButtonHashmap.java
package org.firstinspires.ftc.teamcode.common;

import java.util.HashMap;
import java.util.Map;

public class ButtonHashmap {
    public static ButtonControl map(BunyipsOpMode opMode, String option1, String option2, ButtonControl button1, ButtonControl button2, ButtonControl defaultbutton) {
        HashMap<ButtonControl, String> buttonMap = new HashMap<>();

        buttonMap.put(button1, option1);
        buttonMap.put(button2, option2);
        ButtonControl selectedButton = defaultbutton;

        boolean autoClearState = opMode.telemetry.isAutoClear();

        opMode.telemetry.setAutoClear(false);

        opMode.telemetry.addLine("PRESSING ONE OF THE FOLLOWING BUTTONS WILL INITIALISE THE ROBOT TO THE RELEVANT OPERATION MODE:");
        for (Map.Entry<ButtonControl, String> es : buttonMap.entrySet()) {
            System.out.printf("%s: %s%n", es.getKey().name(), es.getValue());
            opMode.telemetry.addLine(String.format("%s: %s", es.getKey().name(), es.getValue()));
            if (ButtonControl.isSelected(opMode.gamepad1, es.getKey())) {
                selectedButton = es.getKey();
            }
        }

        opMode.telemetry.addLine(String.format("IF NO BUTTON IS PRESSED WITHIN 3 SECONDS, *%s* WILL BE RUN", buttonMap.get(selectedButton)));
        opMode.telemetry.update();


        long startTime = System.nanoTime();

        outerLoop:
        while (System.nanoTime() < startTime + 3000000000L) {
            for (Map.Entry<ButtonControl, String> es : buttonMap.entrySet()) {
                if (ButtonControl.isSelected(opMode.gamepad1, es.getKey())) {
                    selectedButton = es.getKey();
                    break outerLoop;
                }
            }

            opMode.idle();
        }

        opMode.telemetry.addLine(String.format("%s was selected: Running %s", selectedButton.name(), buttonMap.get(selectedButton)));
        opMode.telemetry.addLine("IF THIS SELECTION IS INCORRECT, QUIT THE OPMODE AND CHOOSE RESELECT");
        opMode.telemetry.update();

        opMode.telemetry.setAutoClear(autoClearState);


        opMode.telemetry.addLine(String.format("Loading tasks for %s", selectedButton.name()));
        return selectedButton;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/common/ButtonHashmap.kt
package org.firstinspires.ftc.teamcode.common

import com.qualcomm.robotcore.util.ElapsedTime

/**
 * Convenience object to ask for user input from a controller in order to determine a pre-determined
 * set of instructions before Autonomous.
 *
 * Updated to use dynamic button mapping and generics 04/08/23.
 */
object ButtonHashmap {
    fun <T> map(opMode: BunyipsOpMode, vararg opmodes: T): T {
        if (opmodes.isEmpty()) {
            throw IllegalArgumentException("Must provide at least one operation mode.")
        }

        val buttons: HashMap<T, ButtonControl> = ButtonControl.mapArgs(opmodes)

        // Default button, use the first vararg as the default
        var selectedButton: ButtonControl = buttons.values.first()
        // User selected operation mode
        var selectedOpMode: T? = null

        // Disable auto clear if it is enabled
        val autoClearState = opMode.getTelemetryAutoClear()
        opMode.setTelemetryAutoClear(false)

        // Log available buttons to press
        opMode.addTelemetry("PRESSING ONE OF THE FOLLOWING GAMEPAD1 BUTTONS WILL INITIALISE THE ROBOT TO THE RELEVANT OPERATION MODE:")
        for ((name, button) in buttons) {
            opMode.addTelemetry(String.format("%s: %s", button.name, name))
        }

        opMode.addTelemetry(
            String.format(
                "IF NO BUTTON IS PRESSED WITHIN 3 SECONDS, *%s* WILL BE RUN",
                selectedButton
            )
        )

        // Must manually call telemetry pushes as the BYO is no longer handling it
        opMode.telemetry.update()

        val timeout = ElapsedTime()
        while (timeout.seconds() < 3.0 && selectedOpMode == null && opMode.opModeInInit()) {
            for ((str, button) in buttons) {
                if (ButtonControl.isSelected(opMode.gamepad1, button)) {
                    selectedButton = button
                    selectedOpMode = str
                    break
                }
            }
            opMode.idle()
        }

        opMode.addTelemetry(
            String.format(
                "%s was selected: Running '%s'",
                selectedButton.name,
                selectedOpMode ?: buttons.keys.first()
            )
        )

        opMode.addTelemetry("IF THIS SELECTION IS INCORRECT, QUIT THE OPMODE AND RESELECT NOW")
        opMode.telemetry.update()
        opMode.setTelemetryAutoClear(autoClearState)

        // Default to the first OpMode if none were selected
        return selectedOpMode ?: buttons.keys.first()
    }
}